api/src/main/java/jakarta/servlet/http/HttpServlet.java

```diff
@@ -598,78 +598,135 @@ public void service(ServletRequest req, ServletResponse res) throws ServletExcep
+                if (noBodyOutputStream == null) {
+                    setContentLength(0);
+                } else {
+                    flushWriter();
+
+                    int unCommitted = noBodyOutputStream.getUncommitted();
+                    if (unCommitted <= getBufferSize())
+                        setContentLength(unCommitted);
+                }
+            } else {
+                flushWriter();
+                if (noBodyOutputStream.getUncommitted() >= getSetContentLength())
+                    flushBuffer();
+            }
+        }
+    }
+
+    private void flushWriter() {
+        if (writer != null) {
+            try {
+                // Flush the writer with flushing on the wrapped OutputStream disabled
+                // so that the bytes actually written can be counted.
+                noBodyOutputStream.setFlushable(false);
                 writer.flush();
+            } finally {
+                noBodyOutputStream.setFlushable(true);
             }
-            setContentLength(noBody.getContentLength());
         }
     }
 
     @Override
     public void setContentLength(int len) {
         super.setContentLength(len);
-        didSetContentLength = true;
+        setContentLength = len;
     }
 
     @Override
     public void setContentLengthLong(long len) {
         super.setContentLengthLong(len);
-        didSetContentLength = true;
+        setContentLength = len;
     }
 
     @Override
     public void setHeader(String name, String value) {
         super.setHeader(name, value);
-        checkHeader(name);
+        checkHeader(name, value);
     }
 
     @Override
     public void addHeader(String name, String value) {
         super.addHeader(name, value);
-        checkHeader(name);
+        checkHeader(name, value);
     }
 
     @Override
     public void setIntHeader(String name, int value) {
         super.setIntHeader(name, value);
-        checkHeader(name);
+        checkHeader(name, value);
     }
 
     @Override
     public void addIntHeader(String name, int value) {
         super.addIntHeader(name, value);
-        checkHeader(name);
+        checkHeader(name, value);
+    }
+
+    private void checkHeader(String name, String value) {
+        if ("content-length".equalsIgnoreCase(name)) {
+            setContentLength = Long.valueOf(value);
+        }
     }
 
-    private void checkHeader(String name) {
+    private void checkHeader(String name, int value) {
         if ("content-length".equalsIgnoreCase(name)) {
-            didSetContentLength = true;
+            setContentLength = value;
         }
     }
 
+    @Override
+    public void reset() {
+        super.reset();
+        noBodyOutputStream = null;
+        writer = null;
+        usingOutputStream = false;
+        setContentLength = -1;
+    }
+
+    @Override
+    public void resetBuffer() {
+        super.resetBuffer();
+        flushWriter();
+        setContentLength = -1;
```

Comment:
I don't think this is correct, Javadoc says:

`
Clears the content of the underlying buffer in the response without clearing headers or status code. If the response
      has been committed, this method throws an <code>IllegalStateException</code>.
`

So the only way the content length has been set is if the user explicitly set it, and this method should not mess with headers.